File: CaCompteKit/Sources/DesignSystem/Avatar.swift
import Foundation

/// Doc 08 « Avatars » — trois sources, toutes hors ligne.
public struct Avatar: Sendable, Hashable {
  public enum Kind: Sendable, Hashable {
    case symbol(String)
    case emoji(String)
    case photo(Data)
  }

  public let kind: Kind
  public let palette: PlayerPalette

  public init(kind: Kind, palette: PlayerPalette) {
    self.kind = kind
    self.palette = palette
  }

  /// Sélection curatée d'emoji (animaux, objets, expressions) — seule source d'avatar par
  /// défaut désormais : pas de symbole à choisir, l'emoji vient du hachage du pseudo.
  public static let curatedEmoji: [String] = [
    "🦊", "🐻", "🐼", "🐨", "🐯", "🦁", "🐮", "🐷", "🐸", "🐵",
    "🐔", "🐧", "🐦", "🦆", "🦉", "🦇", "🐺", "🐗", "🐴", "🦄",
    "🐝", "🦋", "🐢", "🐍", "🦎", "🐙", "🦑", "🦀", "🐠", "🐬",
    "🐳", "🦈", "🐌", "🐞", "🦔", "🦥", "🦦", "🦨", "🦡",
    "🎨", "🎮", "🎸", "🚀", "⚽️", "🏀", "🎯", "🎲", "🍕", "🌈",
    "⭐️", "🔥", "💎", "🍀", "🎈",
    "🤖", "👾", "🧙", "🥷", "🏎️", "🌸",
    "🦝", "🐲", "🍄", "⚡", "♥️", "💚", "💙", "🩷",
  ]

  /// Doc 08 : « un joueur créé sans choix reçoit un symbole et une couleur dérivés du hachage
  /// stable de son pseudo » — ici toujours un emoji, jamais un choix manuel. `String.hashValue`
  /// est salé aléatoirement par processus — inutile ici — d'où un FNV-1a manuel, trivialement
  /// portable en Kotlin pour produire le même résultat sur les deux plateformes.
  public static func generated(for nickname: String) -> Avatar {
    let hash = fnv1a(nickname)
    let paletteIndex = Int(hash % 10) + 1
    let emojiIndex = Int((hash / 10) % UInt64(curatedEmoji.count))
    return Avatar(
      kind: .emoji(curatedEmoji[emojiIndex]), palette: PlayerPalette(index: paletteIndex))
  }

  private static func fnv1a(_ string: String) -> UInt64 {
    var hash: UInt64 = 0xcbf2_9ce4_8422_2325
    for byte in string.utf8 {
      hash ^= UInt64(byte)
      hash = hash &* 0x0000_0100_0000_01b3
    }
    return hash
  }
}

File: android/designsystem/src/main/kotlin/com/quimene/designsystem/components/Avatar.kt
package com.quimene.designsystem.components

/** Miroir de `Avatar.Kind` (`Avatar.swift`). */
sealed interface AvatarKind {
    data class Symbol(
        val name: String,
    ) : AvatarKind

    data class Emoji(
        val character: String,
    ) : AvatarKind

    class Photo(
        val data: ByteArray,
    ) : AvatarKind {
        override fun equals(other: Any?): Boolean = other is Photo && data.contentEquals(other.data)

        override fun hashCode(): Int = data.contentHashCode()
    }
}

/** Doc 08 « Avatars » — trois sources, toutes hors ligne. Miroir de `Avatar.swift`. */
data class Avatar(
    val kind: AvatarKind,
    val palette: PlayerPalette,
) {
    companion object {
        /** Sélection curatée d'emoji (animaux, objets, expressions) — seule source d'avatar par
         * défaut : pas de symbole à choisir, l'emoji vient du hachage du pseudo. Liste identique
         * à `Avatar.curatedEmoji` (même ordre — l'index dans cette liste fait partie du calcul
         * déterministe de [generated]). */
        val curatedEmoji: List<String> =
            listOf(
                "🦊",
                "🐻",
                "🐼",
                "🐨",
                "🐯",
                "🦁",
                "🐮",
                "🐷",
                "🐸",
                "🐵",
                "🐔",
                "🐧",
                "🐦",
                "🦆",
                "🦉",
                "🦇",
                "🐺",
                "🐗",
                "🐴",
                "🦄",
                "🐝",
                "🦋",
                "🐢",
                "🐍",
                "🦎",
                "🐙",
                "🦑",
                "🦀",
                "🐠",
                "🐬",
                "🐳",
                "🦈",
                "🐌",
                "🐞",
                "🦔",
                "🦥",
                "🦦",
                "🦨",
                "🦡",
                "🎨",
                "🎮",
                "🎸",
                "🚀",
                "⚽️",
                "🏀",
                "🎯",
                "🎲",
                "🍕",
                "🌈",
                "⭐️",
                "🔥",
                "💎",
                "🍀",
                "🎈",
                "🤖",
                "👾",
                "🧙",
                "🥷",
                "🏎️",
                "🌸",
                "🦝",
                "🐲",
                "🍄",
                "⚡",
                "♥️",
                "💚",
                "💙",
                "🩷",
            )

        /**
         * Avatar déterministe dérivé du pseudo : même pseudo → même avatar, sur les deux
         * plateformes. `String.hashCode()` de la JVM n'est PAS salé aléatoirement (contrairement
         * à `String.hashValue` en Swift) mais diverge de toute façon de l'algorithme Swift —
         * d'où le FNV-1a manuel ci-dessous, bit-exact avec `Avatar.fnv1a` (mêmes constantes,
         * même parcours octet-par-octet UTF-8, même arithmétique 64 bits non signée avec
         * dépassement silencieux).
         */
        fun generated(nickname: String): Avatar {
            val hash = fnv1a(nickname)
            val paletteIndex = (hash % 10uL).toInt() + 1
            val emojiIndex = ((hash / 10uL) % curatedEmoji.size.toULong()).toInt()
            return Avatar(
                kind = AvatarKind.Emoji(curatedEmoji[emojiIndex]),
                palette = PlayerPalette(paletteIndex),
            )
        }

        private fun fnv1a(string: String): ULong {
            var hash = 0xcbf29ce484222325uL
            for (byte in string.toByteArray(Charsets.UTF_8)) {
                hash = hash xor byte.toUByte().toULong()
                hash *= 0x100000001b3uL
            }
            return hash
        }
    }
}
